### ISSUE TITLE
Bug: Inconsistent _fulltestsuite_.luac results

### ISSUE THREAD

**Issue by galibzon** [o3de/o3de-atom-sampleviewer#458]
I'm getting inconsistent test results when running  _fulltestsuite_.luac, and once in a while I get a device removed crash.

1st Run results:
![image](https://user-images.githubusercontent.com/66021303/171739854-5120f94e-e20d-467d-a7ba-6dac6383f5e1.png)

2nd Run results:
![image](https://user-images.githubusercontent.com/66021303/171739953-e3e86cce-71d7-4e70-8383-4d8cc7d18652.png)

3rd Run results:
![image](https://user-images.githubusercontent.com/66021303/171740027-df7e9ab6-33c5-49fc-9235-722ac148c7c3.png)

Once in a while I get this crash:
**Stack Trace**
```
>	Atom_RPI.Private.dll!AZ::Debug::Trace::Crash() Line 229	C++
 	Atom_RPI.Private.dll!AZ::RPI::RPISystemComponent::OnDeviceRemoved(AZ::RHI::Device * device) Line 153	C++
 	[Inline Frame] Atom_RHI_DX12.Private.dll!AZStd::Internal::INVOKE(void(AZ::RHI::RHISystemNotificiationInterface::*)(AZ::RHI::Device *) &) Line 177	C++
 	[Inline Frame] Atom_RHI_DX12.Private.dll!AZStd::invoke(void(AZ::RHI::RHISystemNotificiationInterface::*)(AZ::RHI::Device *) &) Line 55	C++
 	[Inline Frame] Atom_RHI_DX12.Private.dll!AZ::EBusEventProcessingPolicy::Call(void(AZ::RHI::RHISystemNotificiationInterface::*)(AZ::RHI::Device *) &) Line 437	C++
 	Atom_RHI_DX12.Private.dll!AZ::Internal::EBusContainer<AZ::RHI::RHISystemNotificiationInterface,AZ::RHI::RHISystemNotificiationInterface,0,1>::Dispatcher<AZ::EBus<AZ::RHI::RHISystemNotificiationInterface,AZ::RHI::RHISystemNotificiationInterface>>::Broadcast<void (__cdecl AZ::RHI::RHISystemNotificiationInterface::*)(AZ::RHI::Device *),AZ::RHI::Device *>(void(AZ::RHI::RHISystemNotificiationInterface::*)(AZ::RHI::Device *) && func, AZ::RHI::Device * && <args_0>) Line 1360	C++
 	Atom_RHI_DX12.Private.dll!AZ::RHI::Device::SetDeviceRemoved() Line 124	C++
 	Atom_RHI_DX12.Private.dll!AZ::DX12::Device::OnDeviceRemoved() Line 494	C++
 	[External Code]	
```

**Output Error**  
```
The thread 0xa998 has exited with code 0 (0x0).
Automation: 
==================================================================
Automation: Trace::Warning
 D:/GIT/o3de/AtomSampleViewer/Gem/Code/Source/Automation/ScriptReporter.cpp(877): 'void __cdecl AtomSampleViewer::ScriptReporter::ReportScriptWarning(const class AZStd::basic_string<char,struct AZStd::char_traits<char>,class AZStd::allocator> &)'
Automation: Script: Failed to load PNG file 'D:/GIT/o3de/AtomSampleViewer/user/scripts/screenshotslocalbaseline/dx12/MaterialHotReloadTest//15_Red_AfterShaderReload.png' with error 'Cannot open file.'
Automation: ==================================================================
System: > r_ShaderVariantAsyncLoader_ServiceLoopDelayOverride_ms : 0
Automation: Running script 'scripts/msaa_rpi_test.bv.luac'...
Automation: Script: Saving screenshots to D:\GIT\o3de\AtomSampleViewer\user\Scripts\Screenshots\MSAA_RPI\
ImGuiSidebar: Saving settings to 'D:/GIT/o3de/AtomSampleViewer/user/MaterialHotReloadTestComponent/sidebar.xml'
The thread 0xf68c has exited with code 0 (0x0).
ImGuiPass: 
==================================================================
ImGuiPass: Trace::Error
 D:/GIT/o3de/Gems/Atom/Feature/Common/Code/Source/ImGui/ImGuiPass.cpp(88): '__cdecl AZ::Render::ImGuiPass::ImGuiPass(const struct AZ::RPI::PassDescriptor &)'
ImGuiPass: Default ImGui pass is already set on this pipeline, ignoring request to set this pass as default. Only one ImGui pass should be marked as default in the pipeline.
ImGuiPass: ==================================================================
MaterialAsset: 
==================================================================
MaterialAsset: Trace::Warning
 D:/GIT/o3de/Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp(299): 'void __cdecl AZ::RPI::MaterialAsset::ApplyVersionUpdates(class AZStd::function<void __cdecl(char const *)>)'
MaterialAsset: This material is based on version '3' of ['materials/types/standardpbr.azmaterialtype' AssetId={15F89219-4AEA-52A0-A520-A818249665CA}:0 Type={CD7803AB-9C4C-4A33-9A14-7412F1665464}], and the material type is now at version '5'. Automatic updates have been applied. Consider updating the .material source file for {39704C53-AC3A-51BE-81EA-23CEA2455340}:0 (materials/defaultpbr.azmaterial).
MaterialAsset: ==================================================================
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\wer.dll'. 
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\aepic.dll'. 
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\sfc.dll'. 
'AtomSampleViewerStandalone.exe' (Win32): Unloaded 'C:\Windows\System32\sfc.dll'
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\sfc.dll'. 
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\sfc_os.dll'. 
Exception thrown at 0x00007FFCCCDB4FD9 in AtomSampleViewerStandalone.exe: Microsoft C++ exception: std::bad_alloc at memory location 0x000000E19CDFE878.
....
D3D12: Removing Device.
Exception thrown at 0x00007FFCCCDB4FD9 in AtomSampleViewerStandalone.exe: Microsoft C++ exception: _com_error at memory location 0x000000E19CCF7BF0.
Device: Device was removed because of the following reason:
Exception thrown at 0x00007FFCCCDB4FD9 in AtomSampleViewerStandalone.exe: Microsoft C++ exception: _com_error at memory location 0x000000E19CCF8030.
Exception thrown at 0x00007FFCCCDB4FD9 in AtomSampleViewerStandalone.exe: Microsoft C++ exception: _com_error at memory location 0x000000E19CCF8450.
DX12: DXGI_ERROR_DEVICE_HUNG - The application's device failed due to badly formed commands sent by the application. This is an design-time issue that should be investigated and fixed.
Device:  ===========================End of OnDeviceRemoved================================
'AtomSampleViewerStandalone.exe' (Win32): Loaded 'C:\Windows\System32\TextShaping.dll'. 
```

**Comment by galibzon** [comment]
Got the crash again.

1st message box:

**Comment by santorac** [comment]
There are probably separate issues going on here. The one I'm looking into is the MaterialHotReloadTest failures. This appears to be dependent on how many AssetBuilder.exe instances are being run by the AssetProcessor. If you have a fully built cache before starting the AP, there will be only one instance of AssetBuilder.exe and MaterialHotReloadTest will pass. In order for it to fail I have to have many AssetBuilder.exe instances running. 10-15 instances was not enough to make the test fail. I'm not sure what the tipping point is, but my CPU has 64 logical cores and if I have 64(ish) AssetBuilder.exe's running then it will fail. So for me to get a consistent repro, I add some whitespace to StandardPBR_ForwardPass.azsl, wait for the AP to finish processing, and then run MaterialHotReloadTest.

**Comment by santorac** [comment]
Regarding the fulltestsuite crash, I have hit it so far 2 out of 4 attempts. 
It does not seem to be related to the MaterialHotReloadTest failure at all, does not matter if there is only one AssetBuilder.exe.

**Comment by santorac** [comment]
After testing a bunch of different commits, I think the crash issue was introduced by this PR: https://github.com/o3de/o3de/pull/8312

**Comment by santorac** [comment]
DirectX debug layer only indicates that the device was lost as the result of TDR, no other indication of a problem.
```
D3D12: Removing Device.
D3D12 ERROR: ID3D12Device::RemoveDevice: Device removal has been triggered for the following reason (DXGI_ERROR_DEVICE_HUNG: The Device took an unreasonable amount of time to execute its commands, or the hardware crashed/hung. As a result, the TDR (Timeout Detection and Recovery) mechanism has been triggered. The current Device Context was executing commands when the hang occurred. The application may want to respawn and fallback to less aggressive use of the display hardware). [ EXECUTION ERROR #232: DEVICE_REMOVAL_PROCESS_AT_FAULT]
D3D12: **BREAK** enabled for the previous message, which was: [ ERROR EXECUTION #232: DEVICE_REMOVAL_PROCESS_AT_FAULT ]

```

**Comment by santorac** [comment]
This issue might be duplicate of https://github.com/o3de/o3de-atom-sampleviewer/issues/455

**Comment by NeilGitGud** [comment]
Still occurs as of build:

O3de:
commit fa1213139ff6eefc98b35f3a329f727b050a6019

ASV:
commit 8fb8e0a8c1f68d283cca0c07018c2b7fd2596090
